stl/inc/xstring

```diff
@@ -2546,69 +2660,207 @@ public:
+        _Construct<_Construct_strategy::_From_ptr>(_Ptr, _Convert_size<size_type>(_Traits::length(_Ptr)));
     }
 
     _CONSTEXPR20 basic_string(_CRT_GUARDOVERFLOW const size_type _Count, const _Elem _Ch)
-        : _Mypair(_Zero_then_variadic_args_t{}) {
-        // construct from _Count * _Ch
-        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
-        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
-        _Tidy_init();
-        assign(_Count, _Ch);
-        _Proxy._Release();
+        : _Mypair(_Zero_then_variadic_args_t{}) { // construct from _Count * _Ch
+        _Construct<_Construct_strategy::_From_char, _Elem>(_Ch, _Count);
     }
 
 #if _HAS_CXX17
     template <class _Alloc2 = _Alloc, enable_if_t<_Is_allocator<_Alloc2>::value, int> = 0>
 #endif // _HAS_CXX17
     _CONSTEXPR20 basic_string(_CRT_GUARDOVERFLOW const size_type _Count, const _Elem _Ch, const _Alloc& _Al)
         : _Mypair(_One_then_variadic_args_t{}, _Al) { // construct from _Count * _Ch with allocator
-        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
-        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
-        _Tidy_init();
-        assign(_Count, _Ch);
-        _Proxy._Release();
+        _Construct<_Construct_strategy::_From_char, _Elem>(_Ch, _Count);
     }
 
     template <class _Iter, enable_if_t<_Is_iterator_v<_Iter>, int> = 0>
     _CONSTEXPR20 basic_string(_Iter _First, _Iter _Last, const _Alloc& _Al = _Alloc())
         : _Mypair(_One_then_variadic_args_t{}, _Al) {
-        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
-        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
-        _Tidy_init();
         _Adl_verify_range(_First, _Last);
-        auto _UFirst      = _Get_unwrapped(_First);
-        const auto _ULast = _Get_unwrapped(_Last);
-
-        if constexpr (is_same_v<remove_const_t<remove_pointer_t<decltype(_UFirst)>>, _Elem>) {
-            if (_UFirst != _ULast) {
-                assign(_UFirst, _Convert_size<size_type>(static_cast<size_t>(_ULast - _UFirst)));
-            }
+        if (_First == _Last) {
+            _Mypair._Myval2._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal()));
+            _Tidy_init();
         } else {
-            if constexpr (_Is_fwd_iter_v<_Iter>) {
-                const auto _Count = _Convert_size<size_type>(static_cast<size_t>(_STD distance(_UFirst, _ULast)));
-                reserve(_Count);
+            const auto _UFirst = _Get_unwrapped(_First);
+            const auto _ULast  = _Get_unwrapped(_Last);
+            if constexpr (_Is_elem_cptr<decltype(_UFirst)>::value) {
+                _Construct<_Construct_strategy::_From_ptr>(
+                    _UFirst, _Convert_size<size_type>(static_cast<size_t>(_ULast - _UFirst)));
             } else {
-                static_assert(_Is_input_iter_v<_Iter>, "Should be at least input iterator");
+                _Construct_from_iter(_STD move(_UFirst), _STD move(_ULast));

```

Comment:
This moving won't accomplish much because `_UFirst` and `_ULast` are top-level `const`.